### DISCUSSION TITLE
Customizing pagination for ui.table

### DISCUSSION THREAD

**Discussion by ProbablyBrian** [zauberzeug/nicegui#1903] [Q&A] [answered]
### Question

I was looking at the q-table and q-pagination documentation and saw that the pagination element can be configured to use a series of numbered buttons (and prev/next/first/last buttons) instead of the default input-based option.  I was trying to work out how to do this with NiceGUI though all of my attempts have come up short.  

My first thought was to do something similar to:

```python
columns = [...]
rows = [...]

with ui.table(columns=columns, rows=rows, row_key="id", pagination={"rowsPerPage": 2}) as table:
    with table.add_slot("pagination") as pg:
        ui.element("q-pagination").props("max=5") # arbitrary props to get it to appear
```

and while this does get a pagination element to render properly, it doesn't remove or override the existing pagination.  The pagination slot also seems to have no children by default so I'm not able to find out how to remove the existing pagination element.  Also, I'd have no clue how to make sure the new pagination element is bound to or communicating with the parent table object.

I am not proficient with Vue outside of some rudimentary concepts so I might be going about this in a completely incorrect way.  Can anyone help me find a solution that allows me to customize the pagination to use the numbered-button variation of the pagination?

**Comment by falkoschindler** [comment, marked as answer]
After fiddling around with it for quite a while, I came up with this proof of concept:
```py
columns = [
    {'name': 'name', 'label': 'Name', 'field': 'name', 'required': True, 'align': 'left'},
    {'name': 'age', 'label': 'Age', 'field': 'age', 'sortable': True},
]
rows = [
    {'name': 'Alice', 'age': 18},
    {'name': 'Bob', 'age': 21},
    {'name': 'Carol'},
    {'name': 'Dave', 'age': 25},
    {'name': 'Eve', 'age': 30},
    {'name': 'Frank', 'age': 35},
    {'name': 'Grace', 'age': 40},
]

ROWS_PER_PAGE = 3
with ui.table(columns=columns, rows=rows, row_key='name') as table:
    table._props['pagination'] = {'rowsPerPage': ROWS_PER_PAGE}
    with table.add_slot('bottom'):
        ui.pagination(1, int((len(rows) - 1) / ROWS_PER_PAGE) + 1, on_change=lambda e: (
            table.run_method('setPagination', {'rowsPerPage': ROWS_PER_PAGE, 'page': e.value}),
        )).classes('mx-auto')
```

It uses the new `ui.pagination` element which will be released in 1.4.2. In contrast to `ui.element('q-pagination')` it simplifies registering to change events.

And by setting `_props['pagination']` instead of using the `pagination` parameter of `ui.table`, it doesn't show the "pagination" slot. Instead we add the pagination element to the "bottom" slot.

But it all gets more complicated when the data changes. You need to keep track of the current page and update the pagination if the number of rows changes.

**Comment by ProbablyBrian** [comment]
That looks great!  Keeping track of the pages isn't such a big deal; I'm already managing the filters/columns/DB queries to populate the table so I should have the information necessary.  

On a slightly unrelated note... Any plans to implement the capability for using server-side loading for tables?

**Reply by falkoschindler** [reply]
No, we haven't thought about server-side loading for tables. Could be interesting for large data sets. But I'm not sure how easy it would be to implement.

**Comment by bmaranville** [comment]
Here is how I am doing server-side pagination for ui.table.  By setting `pagination['rowsNumber']` you signal Quasar `q-table` to use server-side pagination, then you listen to the `request` event and pull new rows from the API based on that, replacing the existing ones.  Finally call `refresh()` on the refreshable table element.

```python
from nicegui import ui
import requests

API = 'https://some.api'

table_data = {
    'rows': [],
    'pagination': {
        'rowsPerPage': 10,
        'descending': True,
        'sortBy': 'date',
        'page': 0,
        'rowsNumber': 0,
    },
}

@ui.refreshable
def paginated_table():
    columns = [
        {'name': 'name', 'label': 'Name', 'field': 'name', 'required': True, 'align': 'left'},
        {'name': 'age', 'label': 'Age', 'field': 'age', 'sortable': True},
    ]
    table = ui.table(
        columns=columns,
        rows=table_data['rows'],
        row_key='name',
        pagination=table_data['pagination'],
    )
    table.on('request', do_pagination)

def get_full_count():
    full_count = requests.get(API, params={'full_count': True}).json()
    table_data['pagination']['rowsNumber'] = full_count
    paginated_table.refresh()

def do_pagination(request):
    new_pagination = request.args['pagination']
    pagination = table_data['pagination']
    pagination.update(new_pagination)
    # this assumes your API can take "page" and "rowsPerPage" parameters...
    # you might need to translate to "offset" and "limit"
    new_rows = requests.get(API, params=pagination).json()
    table_data['rows'] = new_rows
    paginated_table.refresh()

ui.page_title("Paginated Table")
paginated_table()
get_full_count()
```
